permit2.ts

```diff
@@ -1,52 +1,74 @@
-import { ethers, BigNumber, ContractTransaction } from 'ethers';
-import { Contract as MulticallContract, Provider as MulticallProvider } from 'ethers-multicall';
+import { BigNumber, ContractTransaction } from 'ethers'
+import { Contract as MulticallContract, Provider as MulticallProvider } from 'ethers-multicall'
 import Permit2Abi from '@/abis/Permit2.json'
-import { Permit2__factory } from '@/types/ethers-contracts';
-import {
-  PermitBatch,
-  PermitBatchData,
-  AllowanceTransfer,
-  MaxSigDeadline
-} from '@uniswap/permit2-sdk'
+import { Permit2__factory } from '@/types/ethers-contracts'
+import { PERMIT2_ADDRESS } from '@uniswap/permit2-sdk'
+import { Provider, Signer } from '@wagmi/core'
 
-const PERMIT2="0x000000000022d473030f116ddee9f6b43ac78ba3"
+type Token = string
+type Spender = string
 
 export type TokenSpender = {
-  token: string;
-  spender: string;
+  token: Token
+  spender: Spender
 }
 
 export type Allowances = {
-  [token: string]: {
-    [spender: string]: {
-      amount: BigNumber;
-      expiration: number;
+  [spender: string]: {
+    [token: string]: {
+      amount: BigNumber
+      expiration: number
     }
   }
 }
 
-export async function useAllowances(provider: ethers.providers.Provider, owner: string, tokenSpenders: TokenSpender[]): Promise<Allowances> {
-  const multicallProvider = new MulticallProvider(provider);
-  await multicallProvider.init();
-  const permit2Contract = new MulticallContract(PERMIT2, Permit2Abi);
+// gets all tokens and spenders ever permitted
+export async function getPermittedTokensAndSpenders(provider: Provider, address: string): Promise<TokenSpender[]> {
+  const permit2Contract = Permit2__factory.connect(PERMIT2_ADDRESS, provider)
+  const filter = permit2Contract.filters.Permit(address)
+  const events = await permit2Contract.queryFilter(filter)
+
+  // get unique tokens and spenders
+  const result = []
+  for (const event of events) {
+    const args = event.args
+    const token = args?.[1]
+    const spender = args?.[2]
+    result.push({ token, spender })
+  }
+
+  return result
+}
+
+// gets all tokens and spenders and amounts currently permitted
+export async function getAllowances(provider: Provider, address: string, tokenSpenders: TokenSpender[]): Promise<Allowances> {
+  const multicallProvider = new MulticallProvider(provider)
+  await multicallProvider.init()
+  const permit2Contract = new MulticallContract(PERMIT2_ADDRESS, Permit2Abi)
 
   const calls = tokenSpenders.map(({ token, spender }) =>
-    permit2Contract.allowance(owner, token, spender)
-  );
+    permit2Contract.allowance(address, token, spender)
+  )
 
-  const results = await multicallProvider.all(calls);
+  const results = await multicallProvider.all(calls)
 
   return results.reduce((allowances, [amount, expiration], i) => {
-    const { token, spender } = tokenSpenders[i];
-    if (!allowances[token]) {
-      allowances[token] = {};
+    const { token, spender } = tokenSpenders[i]
```

Comment:
dis kinda messy

can it be a for loop instead xD

Reply:
waa you dont like functional style?

Reply:
bruhhh